### DISCUSSION TITLE
Why is getServerSideProps async and how can i have it behave like an async function?

### DISCUSSION THREAD

**Discussion by laurencefass** [vercel/next.js#37717] [Help]
I wish to circumvent the white screen appearing on any page delayed by any form of delayed data fetch in getServerSideProps and I dont want to resort to pulling data client site as a workaround. getServerSideProps acts in a synchronous blocking manner from the perspective of my application. i.e. if an asynchronous read 

its signature is:
```
export async function getServerSideProps(context) {
  return {
    props: {}, // will be passed to the page component as props
  }
}
```
so its async but it acts synchronously from the perspective of my application and blocks any blocking ops I perform within it.

as an experiment (doomed to fail) I tried:

```
export async function getServerSideProps({ req, res }) {
    return {
        props: {
            promise: new Promise((resolve, reject) => {
                setTimeout(() => {
                    resolve('foo');
                }, 300);
            })
        },
    }
}
``` 

but I got an error:

```
Server Error
Error: Error serializing `.promise` returned from `getServerSideProps` in "/promise".
Reason: `object` ("[object Promise]") cannot be serialized as JSON. Please only return JSON serializable data types.
```

So

1. Why is getServerSideProps async when it behaves like a synchronous function and wont allow return of a promise?
2. is there a way of using getServerSideProps in an asynchronous way for example to display a loader while waiting for a serverside fetch to complete?

Thanks

**Comment by icyJoseph** [comment]
Hi,

`getServerSideProps` can be both, sync or async, it just depend son how you define it.


### Sync Mode

```ts
// sync mode
export const getServerSideProps = (context) => {
  return {props: {}};
}
```

### Async Mode

```ts
// async mode
export const getServerSideProps = async (context) => {
  return {props: {}};
}
```

Same goes for regular `function` keyword definitions. Do remember that `async` is a rather modern concept, and in old times, people used callbacks to instruct when something asynchronous was completed.

### Your example

You cannot send instances, or rather, you can only send serializable data through `getServerSideProps`, most specifically, the `props` have to be `JSON.stringify` friendly. That means, that self referential objects, BigInt, Dates, Sets, Maps, etc need to somehow serialized when returned as part of props. 

In your code you need to `await` the promise.

```ts
export async function getServerSideProps({ req, res }) {
    return {
        props: {
           // await this promise!
            promise: await new Promise((resolve, reject) => {
                setTimeout(() => {
                    resolve('foo');
                }, 300);
            })
        },
    }
}
```

### Questions

> 1. Why is getServerSideProps async when it behaves like a synchronous function and wont allow return of a promise?

It can be either, I think you might be confused with what's asynchronous work on JavaScript.

The problem here is that the result of `getServerSideProps` must be serializable. 

A silly analogy would be to think that, the result of `getServerSideProps` should be stateless, and de-hydratable, so that the client can hydrate it again, at any time. 

In other words, you must be able to stringify it, and then parse it into an object again. 

> 2. is there a way of using getServerSideProps in an asynchronous way for example to display a loader while waiting for a server side fetch to complete?

The reason why we use the `async` mode in `getServerSideProps` is because some transactions against other APIs are asynchronous, and we'd like to avoid callback-hell, and instead, we just await for each transaction and keep on building the final result, which must be serializable.

Because pages that use `getServerSideProps` are rendered on demand, there's two cases to handle. The user directly goes to the page, in which case, the request arrives to the server, `gSSP` is called, its result collected, and used to render the page on the server, then next.js serves the HTML for this page, which is then hydrated by the client.

The other case, where the user is already in your app, and they navigate to your page, as an SPA transition, then Next.js pre-fetches the JavaScript that runs the page, and as soon as the user clicks on the link to navigate, Next.js fetches a JSON file, which is the result of running `gSSP` on the server (image trying to pass promise/class instance over HTTP), to use it as props for the page to render the page on the client.

One thing you could do is to listen for `Router.events.on("routeChangeStart", /* start a spinner if the url matches your gSSP page*/)`, and then inside a `useEffect` on the page that uses `gSSP`, run a function to kill the spinner. 

You'd have to tweak this until it works the way you want it. I'd use the `useEffect`, because then I know the client has rendered the page. I am a bit busy these days so I can't really produce a code example for you.

**Reply by kopach** [reply]
Based on type definition, `getServerSideProps` should always return promise
https://github.com/vercel/next.js/blob/6bc07792a4462a4bf921a72ab30dc4ab2c4e1bda/packages/next/types/index.d.ts#L302
Is this a defect in type definition then?

**Reply by icyJoseph** [reply]
Not necessarily.

You can perfectly have:

```tsx
import Header from '../components/header';

export default function IndexPage(props: unknown) {
  console.log(props);
  return (
    <>
      <Header />
      <h1>Hello World!</h1>
    </>
  );
}

export const getServerSideProps = () => {
  return { props: { foo: 4 } };
};

```

And that'll work, because you can always `await` a function call, even if it is not a promise. In general, you can `await` expressions. 

What gives then?

If you want to type the function using next's `GetServerSideProps` type, then you'd find it complains about the promise:

```jsx
import Header from '../components/header';
import type { GetServerSideProps } from 'next';

export default function IndexPage(props: unknown) {
  console.log(props);
  return (
    <>
      <Header />
      <h1>Hello World!</h1>
    </>
  );
}

// type error - but it works
export const getServerSideProps: GetServerSideProps = () => {
  return { props: { foo: 4 } };
};

```

One could argue that maybe it should be expanded to:

```ts
export type GetServerSideProps<
  Props extends { [key: string]: any } = { [key: string]: any },
  Params extends ParsedUrlQuery = ParsedUrlQuery,
  Preview extends PreviewData = PreviewData
> = (
  context: GetServerSidePropsContext<Params, Preview>
) => GetServerSidePropsResult<Props> | Promise<GetServerSidePropsResult<Props>>
```

But, the vast majority of GSSP use cases, end up requiring async work within, and you could still type the arguments (with GetServerSidePropsContext), or the return type (with GetServerSidePropsResult). 

"Returning a promise" is also rather nuanced, it is sufficient to tag the function as `async`, you don't necessary have to have a promise in the return statement.

**Comment by laurencefass** [comment]
Thanks!

I knew returning the promise would fail I just wanted to see it in action. It was my own wishful thinking to envision that the promise result could potentially be serialised across the network internally by Next. Not yet I think.

The problem I am experiencing using gssp is that the entire page is delayed waiting for data fetching or awaiting other async/promises to resolve.  I think you described it well as:

" The user directly goes to the page, in which case, the request arrives to the server, gSSP is called, its result collected, and used to render the page on the server, then next.js serves the HTML for this page, which is then hydrated by the client."

This has always felt like a bad UX to me and the only way I can see to decrease TTFB is to useEffect.fetch the data from the component and not use gssp, but this feels like a workaround for a platform which boasts SSR as a primary feature.

I think that React 18 concurrency & data streaming may fix this sometime in the near future (Ive raised another discussion somewhere else in here and apparently Remix has pipelined exactly this functionality) but Im still exploring whether there are ways around this today. i.e. to use gssp and to load a skeleton while I wait for the data to load. I guess not (for now, at least...).

I think there is some potential for gssp and Suspense to work together but I havent seen any active discussion on this point so any links are appreciated.

**Reply by icyJoseph** [reply]
Well, you got that right, this discussion from the react working group, https://github.com/reactwg/react-18/discussions/37, pretty much outlines the issue, either we pre-render a build time, or we ought to wait for the data on every request as needed and stream it all at once. I am paraphrasing here.

This is why `gSSP` is also a bit of a niche use case, ISR and client-side-fetching with api routes can achieve almost everything else.

> You should use getServerSideProps only if you need to render a page whose data must be fetched at request time. This could be due to the nature of the data or properties of the request (such as authorization headers or geo location). Pages using getServerSideProps will be server side rendered at request time and only be cached if [cache-control headers are configured](https://nextjs.org/docs/going-to-production#caching).
>
> If you do not need to render the data during the request, then you should consider fetching data on the [client side](https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props#fetching-data-on-the-client-side) or [getStaticProps](https://nextjs.org/docs/basic-features/data-fetching/get-static-props).

So it is not like there's zero ways out of it. ISR with short-time/on-demand revalidation can be just as powerful.

**Comment by laurencefass** [comment]
Thanks for your feedback and opinion. One more question if I may to conclude this discussion: Do you happen to know if Next JS will integrate gssp and Suspense in the future to eliminate the white screen waiting while waiting for data to load in the backend when fetching through gssp? Or perhaps provide some other SSR solution?

**Reply by icyJoseph** [reply]
Most definitely! They are working on it. 

We are in the early days of suspense still, and there are also server components in the works.

Suspense as it is, is not necessarily the ultimate solution, but rather part of a bigger scheme. The future looks bright though :)

 - https://gist.github.com/shuding/6ef6a85c4c8ee57d9926e705adef88e3
- https://github.com/vercel/swr/issues/1898

**Comment by laurencefass** [comment]
All of IcyJospeh's replies above constitute a comprehensive answer to the current state of Next JS getServerSideProps. Thankyou!

**Reply by icyJoseph** [reply]
Nice :D please do consider marking as answered if you will! It is always nice to explore this concepts and see what has changed since the last time.
